### PR TITLE
See through ref pointer indirection when optimizing stack locals

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn/issues/53113. Previously, when we did stack local optimization, we'd erase the assignment (including the context of the assignment being by-ref), leaving just the pointer indirection. This is incorrect, and resulted in us generating a ldind where we shouldn't. This fixes the bug by tracking whether we're currently optimizing a stack-scheduled assignment, and seeing through the pointer indirection in that case.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs:1219]: >UnsafeReleaseExe [](http://example.com/codeflow?start=68&length=16)

We should test this case with DEBUG as well.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefLocalTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefLocalTests.cs:3427]: >UnsafeReleaseExe [](http://example.com/codeflow?start=68&length=16)

We should test with DEBUG as well.

**Comment by cston** [comment]
>             if (isLast)

Shouldn't we reset `_isStackScheduledRefAssignment` here?

---
Refers to: src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs:2096 in cfd9cf5. [](commit_id = cfd9cf5dca8e2366cb76bc609c275fc48bf441fc, deletion_comment = False)

**Comment by 333fred** [comment]
@cston thanks for the initial review. I've updated the implementation a bit, let me know if you want walk through it at all.

File: src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs:2090]: >false [](http://example.com/codeflow?start=45&length=5)

Should this reset to the previous value rather than `false`? (Is there a test we can write where that makes a difference?)

Comment (333fred) [reply src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs:2090]: No, I asserted that it was false above as well. `Visit` should be saving and resetting the value unless the thing being visited is an indirection or conditional operator, so we should never be observing a non-false value here.

Comment (cston) [reply src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs:2090]: Ok, thanks.

Could we save and restore the value anyway, for clarity, and to avoid a bug later if there are additional cases where the save and restore in `Visit` is not sufficient?

Comment (333fred) [reply src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs:2090]: I suppose I could, but I'll keep the assertion anyway. If we come across a case like that we definitely need to understand the semantics more completely.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs:1266]: I think we typically use `Verification.Skipped` when valid code is not handled by the verifier. `Verification.Fails` seems to imply a bug.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs:1266]: Other ref local code used `Fails`, which is why I defaulted to that here.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:345]: > [](http://example.com/codeflow?start=4&length=89)

Consider including information about how to interpret the values.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:347]: I am not sure why do we need to calculate this value in Optimizer and limit that just to an assignment context. It feels like we can determine if a BoundPointerIndirectionOperator node is actually a dereference during an initial binding. If it is a target of a ```ref``` expression, it is not a dereference. If we capture this information during initial binding, the implementation wouldn't need to worry about other `ref` propagating scenarios like conditional expression, switch expression, etc.

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:680]: >EmitAddress(expression, AddressKind.Writeable); [](http://example.com/codeflow?start=16&length=47)

It looks like we can simply do ```EmitExpression(expression.Operand, used: true);``` and share that across both branches.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 5)

**Comment by alrz** [comment]
I imagine this will also avoids the local being stacked in a ref assignment like this: `outParam1 = outParam2 = null;`, is that correct?

**Comment by 333fred** [comment]
> I imagine this will also avoid the local being stacked in a ref assignment like this: `outParam1 = outParam2 = null;`, is that correct?

I'm not sure what you're asking @alrz: that isn't a ref assignment, this code should have no impact on it.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (333fred) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:349]: I like this refactoring, and went ahead and implemented it.

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback

---
In reply to: [932432493](https://github.com/dotnet/roslyn/pull/56790#issuecomment-932432493) [](http://example.com/codeflow?ancestors=932432493,932432493)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:346]:
```suggestion
         will emit the address of the target, and not actually perform an indirect load (in practice just
```

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:358]:
```suggestion
         will emit the address of the target, and not actually perform an indirect load (
         in in practice just
```

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:348]: >IsAssignmentAddress [](http://example.com/codeflow?start=17&length=19)

Do we need the "Assignment" part in the name? Same for the other node.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:346]: >in in [](http://example.com/codeflow?start=89&length=5)

Double "in"

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:281]: >isAssignmentAddress [](http://example.com/codeflow?start=85&length=19)

It feels like "onlyRefersToLocation" is a better name.

File: src/Compilers/CSharp/Portable/BoundTree/Constructors.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Constructors.cs:731]: >BoundPointerIndirectionOperator [](http://example.com/codeflow?start=15&length=31)

I think adding constructors like that is error prone, easy to miss the fact that a "non-default" value should be used. It also doesn't look like constructors are used very often, not that many changes are avoided. Same comment for the other bound node.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 7)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [932585188](https://github.com/dotnet/roslyn/pull/56790#issuecomment-932585188) [](http://example.com/codeflow?ancestors=932585188,932585188)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 8), assuming CI is passing.

**Pull request merged**
